src/main/kotlin/io/sip3/salto/ce/sdp/SdpHandler.kt

```diff
@@ -0,0 +1,175 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.sip3.salto.ce.sdp
+
+import gov.nist.javax.sip.message.SIPMessage
+import gov.nist.javax.sip.message.SIPRequest
+import gov.nist.javax.sip.message.SIPResponse
+import io.sip3.commons.domain.SdpSession
+import io.sip3.commons.util.IpUtil
+import io.sip3.salto.ce.RoutesCE
+import io.sip3.salto.ce.USE_LOCAL_CODEC
+import io.sip3.salto.ce.util.callId
+import io.sip3.salto.ce.util.matches
+import io.vertx.core.json.JsonObject
+import io.vertx.kotlin.coroutines.CoroutineVerticle
+import mu.KotlinLogging
+import org.restcomm.media.sdp.SessionDescription
+import org.restcomm.media.sdp.SessionDescriptionParser
+import org.restcomm.media.sdp.attributes.RtpMapAttribute
+import org.restcomm.media.sdp.fields.MediaDescriptionField
+
+class SdpHandler : CoroutineVerticle() {
+
+    private val logger = KotlinLogging.logger {}
+
+    private var codecs = mapOf<String, Codec>()
+    private val sessions = mutableMapOf<String, SdpSessionDescription>()
+
+    override suspend fun start() {
+        readCodecs(config)
+        vertx.eventBus().localConsumer<JsonObject>(RoutesCE.config_change) { event ->
+            try {
+                val config = event.body()
+                readCodecs(config)
+            } catch (e: Exception) {
+                logger.error("SdpHandler `readCodecs()` failed.", e)
+            }
+        }
+
+        vertx.eventBus().localConsumer<SIPMessage>(RoutesCE.sdp_session_update) { event ->
+            try {
+                val message = event.body()
+                handleSdp(message)
+            } catch (e: Exception) {
+                logger.error("SdpHandler 'handleSdp()' failed.", e)
+            }
+        }
+
+        vertx.eventBus().localConsumer<String>(RoutesCE.sdp_session_terminate) { event ->
+            try {
+                val sessionKey = event.body()
+                sessions.remove(sessionKey)
+            } catch (e: Exception) {
+                logger.error("SdpHandler sdpSession remove failed.", e)
+            }
+        }
+    }
+
+    private fun readCodecs(config: JsonObject) {
+        val tmpCodecs = mutableMapOf<String, Codec>()
+        config.getJsonArray("codecs")?.forEach { codecObject ->
+            val codec = (codecObject as JsonObject).mapTo(Codec::class.java)
+            tmpCodecs[codec.name!!] = codec
+        }
+        codecs = tmpCodecs
+    }
+
+    private fun handleSdp(message: SIPMessage) {
+        val sessionDescription: SessionDescription = getSessionDescription(message) ?: return
+
+        val callId = message.callId()!!
+        sessions.getOrPut(callId) { SdpSessionDescription(callId) }.apply {
```

Comment:
I think to be clear here we may use transactionId as an SDP session identificator - it defines SDP session fully